---
title: "TU_selection"
author: "Fabian Blombach"
date: "17/03/2019"
output: html_document
---
This script selects transcription units (TUs) with the following criteria for two biological replicates:  
- mapped TSS data from Wurtzel et al. (2009) are used when available, otherwise the start codon is used as proxy for the TSS  
- assigned to TFB and TFEb peaks with biunique correspondence  
- minimum length of 100 bp   
- a minimum coverage of 20 reads in the chromatin input files for the relevant region  
- no TU-internal TFB peaks indicating gene-internal or anti-sense promoters  

The deeptools package needs to be installed!

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
rm(list=ls())

# list of connections to external files
#ChIP-seq peak calling data
TFB_peaks <-  "data/TFB1.expon.peaksIDR0.05.bed"
TFEb_peaks <- "data/TFEb.expon.peaksIDR0.05.bed"

#genome annotations
rRNA_TSS <- "genome_data/rRNA_TSS.bed"
primTSS <- "genome_data//primaryTSS_Wurtzel_uniqueTUassigned.bed"
startCodons <- "genome_data/GenesStartsTUsWithoutTSS.bed"
operons<-read.delim("genome_data/Wurtzel_allTUs.txt", header=T, sep="\t")

#peak calling data for TFB non-template strand ChIP-exo
TSS.adjust<-read.table("data/adjust.TSS.txt", header=T)
TSS.adjust$operon_id<-as.character(TSS.adjust$operon_id)
TSS.adjust$shift<-as.numeric(as.character(TSS.adjust$shift))


#ChIP-seq data
#data available at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE141286
#path to directory where data are stored
mydir<-("../GSE141290/")
inputCov_r1 <-paste0(mydir, "GSM4200383_input.expon_r1.Coverage.bw")
inputCov_r2 <-paste0(mydir, "GSM4200384_input.expon_r2.Coverage.bw")

phase = "expon"

Sys.setenv(phase = phase,
           TFB_peaks = TFB_peaks, 
           TFEb_peaks = TFEb_peaks, 
           rRNA_TSS = rRNA_TSS, 
           primTSS = primTSS,
           startCodons = startCodons,
           inputCov_r1 = inputCov_r1, 
           inputCov_r2 = inputCov_r2)

```

```{bash matching TFB and TFEb peaks to TSSs}
export PATH=$PATH:/Users/fabianblombach/miniconda3/bin
mkdir temp
#merge TFB and TFEb peaks
bedtools window -w 40 -a $TFB_peaks -b $TFEb_peaks > temp/btWindow40.TFB1.TFEb."$phase"_IDR0.05.bed

#assign TFB/TFEb peaks to TSS and start codons
bedtools window -w 40 -a temp/btWindow40.TFB1.TFEb."$phase"_IDR0.05.bed -b $startCodons > temp/btWindow40.TFB1.TFEb."$phase".startCodonsTUsWithoutTSS.bed

bedtools window -w 40 -a temp/btWindow40.TFB1.TFEb."$phase"_IDR0.05.bed -b $primTSS > temp/btWindow40.TFB1.TFEb."$phase".primTSS_uniqueAssigned.bed

bedtools window -w 40 -a temp/btWindow40.TFB1.TFEb."$phase"_IDR0.05.bed -b $rRNA_TSS > temp/btWindow40.TFB1.TFEb."$phase".rRNA_TSS.bed
```


```{r filter for biunique assignment of peaks to TSSs}
# control for biunique assigment of peaks to each other
peaks<-read.table(paste0("temp/btWindow40.TFB1.TFEb.",phase,"_IDR0.05.bed"))
nonuniqueTFB<- peaks[duplicated(peaks$V4),]
nonuniqueTFEb<- peaks[duplicated(peaks$V9),]

data.AUG<- read.table(paste0("temp/btWindow40.TFB1.TFEb.", phase, ".startCodonsTUsWithoutTSS.bed"))
data.TSS <- read.table(paste0("temp/btWindow40.TFB1.TFEb.", phase, ".primTSS_uniqueAssigned.bed"))
data.rRNA<- read.table(paste0("temp/btWindow40.TFB1.TFEb.", phase, ".rRNA_TSS.bed"))
data.rRNA$V14<-c("ID_669")
data.AUG$V14<-sub("operon", "ID", data.AUG$V14)
data.all<- rbind(data.AUG, data.TSS, data.rRNA)
# NOTE: "invalid factor level, NA generated" message for score column (5th) for TSS, ignore!
data.u<- data.all[as.numeric(setdiff(row.names(data.all), row.names(subset(
  data.all, duplicated(V4, fromLast=F)| duplicated(V4, fromLast=T))))),]
data.u<-data.u[order(data.u$V12),]
data.u$V15 <- rep(".", nrow(data.u))
#filter any nonuniquely paired TFB or TFEb peaks from data.u
data.u<-subset(data.u,  !(V4 %in% nonuniqueTFB$V4)  & !(V9 %in% nonuniqueTFEb$V9))

# calculate TU length for each operon ID, based on Wurtzel data for gene start, end and primary TSS and Operon ID assignment
opCoord<- data.frame(operon_ID=c(paste0(rep("ID_", 2232), 1:2232)), start=(rep(0, 2232)), end=(rep(0,2232))) 
for (i in 1:2232) { start<- min(subset(operons, operon_ID == i)[,c(1,8)], na.rm=T);
end<-max(subset(operons, operon_ID == i)[,c(2,8)], na.rm=T);
opCoord[i,2]<-start;
opCoord[i,3]<-end
}
opCoord$length<- opCoord$end - opCoord$start
data.u<-merge(data.u, opCoord[,c(1,4)], by.x="V14", by.y="operon_ID")
#check number of TUs above minimum length
nrow(subset(data.u, length>=100))
#export bed file with coord for all TSS and start codons to calculate coverage data
write.table(data.u[,c(12:14,1,15:16)], paste0("temp/TSSvATG.uniqueTFBvTFEbOccup.", phase, ".bed"), sep="\t", row.names=F, col.names=F, quote=F)
# Count TUs that pass unique peak assignment
nrows<-nrow(data.u)
print(nrows)
print(nrow(data.u))
Sys.setenv(nrows = nrows)
```

Calculate and import coverage data for -100 to +100 interval from deeptools computeMatrix reference-point -bs 1 for both replicates
```{bash calculating chip-seq coverage data}
export PATH=$PATH:/Users/fabianblombach/miniconda3/bin

#calculate input coverage
computeMatrix reference-point --sortRegions keep -bs 1 -a 100 -b 100 -R temp/TSSvATG.uniqueTFBvTFEbOccup."$phase".bed -S $inputCov_r1 $inputCov_r2  -o temp/trashme.npz --outFileNameMatrix temp/trashme.tab;
rm temp/trashme.npz;
head -n 2 temp/trashme.tab > temp/inputCoverage."$phase".tab; 
tail -n $nrows temp/trashme.tab >> temp/inputCoverage."$phase".tab; rm temp/trashme.tab


#count TFB peaks in TU
bedtools window -l 40  -r 100 -sw -c  -a temp/TSSvATG.uniqueTFBvTFEbOccup."$phase".bed -b $TFB_peaks > temp/TSSvATG.uniqueTFBvTFEbOccup.TU500."$phase".intTFB1PeakCount.bed
```

```{r}
cov <- read.table(paste0("temp/inputCoverage.", phase, ".tab"), sep="\t", header=F)
TFBcount<- read.table(paste0("temp/TSSvATG.uniqueTFBvTFEbOccup.TU500.", phase, ".intTFB1PeakCount.bed"), sep="\t", header=F)
data.u$r1.minCovFull <- do.call(pmin, cov[,1:200])
data.u$r2.minCovFull <- do.call(pmin, cov[,201:400])
```
fraction of TUs > 100 bp
```{r}
print(nrow(subset(data.u, length >=100))/nrows)
```
fraction of TUs with coverage min >20 bp
```{r}
print(nrow(subset(data.u, r1.minCovFull >20 & r2.minCovFull >20))/nrows)
```
fraction of TUs with no internal TFB peaks
```{r}
print(nrow(data.u[TFBcount$V7 == 1,])/nrows)
```
filter dataset for TUs matching all criteria
```{r}
data.u100f<-subset(data.u,r1.minCovFull >20 & 
                          r2.minCovFull >20 &
                          length >=100 &     
                          TFBcount$V7 == 1)
```

fraction of TUs that passes filtering
```{r}
print(nrow(data.u100f)/nrows)
```

number of TUs that passes filtering
```{r}
print(nrow(data.u100f))
```

```{r data export}
write.table(data.u100f[order(data.u100f$V5, decreasing = T),c(12:14,1,15:16)], paste0("TSSvATG.uniqueTFBvTFEbOccup.TU100.intTFBfilt.", phase,".bed"), sep="\t", row.names=F, col.names=F, quote=F)
```

```{r subsets with mapped TSS}
# creating bed files with subsets of the final data set with mapped TSS (TSS & rRNA promoters)
data.TSSrRNA.u100f<-subset(data.u100f, V14 %in% setdiff(data.u100f$V14, data.AUG$V14))
#reorder based on TFB occupancy
data.TSSrRNA.u100f <- data.TSSrRNA.u100f[order(data.TSSrRNA.u100f$V5, decreasing = T),]
write.table(data.TSSrRNA.u100f[order(data.TSSrRNA.u100f$V5, decreasing = T),c(12:14,1,15:16)], paste0("TSS.uniqueTFBvTFEbOccup.TU100.intTFBfilt.", phase,".bed"), sep="\t", row.names=F, col.names=F, quote=F)
```
number of TUs that passes filtering
```{r}
print(nrow(data.TSSrRNA.u100f))
```


```{bash}
rm -r temp
```